File: py/codesamples/list_entities.py
'''
Copyright 2023 Aalyria Technologies, Inc., and its affiliates.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
'''

import grpc
import sys
from pathlib import Path

import api.nbi.v1alpha.nbi_pb2 as Nbi
import api.nbi.v1alpha.nbi_pb2_grpc as NetOpsGrpc
from py.authentication.spacetime_call_credentials import SpacetimeCallCredentials


def main():
    if len(sys.argv) < 4:
        print("Error parsing arguments. Provide a value for host, " + 
                "agent email, agent private key ID, and agent private key file.",
              file=sys.stderr)
        sys.exit(-1)

    HOST = sys.argv[1]
    AGENT_EMAIL = sys.argv[2]
    AGENT_PRIV_KEY_ID = sys.argv[3]
    AGENT_PRIV_KEY_FILE = sys.argv[4]
    PORT = 443

    # The private key should start with "-----BEGIN RSA PRIVATE KEY-----" and
    # end with "-----END RSA PRIVATE KEY-----". In between, there should be newline-delimited
    # strings of characters.
    private_key = Path(AGENT_PRIV_KEY_FILE).read_text()

    # Sets up the channel using the two signed JWTs for RPCs to the NBI.
    credentials = grpc.metadata_call_credentials(
        SpacetimeCallCredentials.create_from_private_key(
            HOST, AGENT_EMAIL, AGENT_PRIV_KEY_ID, private_key))
    channel = grpc.secure_channel(
        f"{HOST}:{PORT}",
        grpc.composite_channel_credentials(grpc.ssl_channel_credentials(),
                                           credentials))

    # Sets up a stub to invoke RPCs against the NBI's NetOps service.
    stub = NetOpsGrpc.NetOpsStub(channel)

    # This stub can now be used to call any method in the NetOps service.
    request = Nbi.ListEntitiesRequest(type="PLATFORM_DEFINITION")
    entities = stub.ListEntities(request)
    print("ListEntitiesResponse received:\n", entities)


if __name__ == "__main__":
    main()

File: java/com/aalyria/spacetime/codesamples/nbi/client/ListEntities.java
// Copyright 2023 Aalyria Technologies, Inc., and its affiliates.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//	http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.aalyria.spacetime.codesamples.nbi.client;

import static io.grpc.Grpc.newChannelBuilderForAddress;

import com.aalyria.spacetime.api.nbi.v1alpha.Nbi.EntityType;
import com.aalyria.spacetime.api.nbi.v1alpha.Nbi.ListEntitiesRequest;
import com.aalyria.spacetime.api.nbi.v1alpha.Nbi.ListEntitiesResponse;
import com.aalyria.spacetime.api.nbi.v1alpha.NetOpsGrpc;
import com.aalyria.spacetime.authentication.SpacetimeCallCredentials;
import io.grpc.CompositeChannelCredentials;
import io.grpc.ManagedChannel;
import io.grpc.TlsChannelCredentials;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;

/**
 * A sample class that sets up the authentication parameters to call Spacetime's Northbound
 * Interface (NBI).
 */
public class ListEntities {
  private static final int PORT = 443;

  public static void main(String[] args) {
    if (args.length < 4) {
      System.err.println(
          "Error parsing arguments. Provide a value for host, "
              + "agent email, agent private key ID, and agent private key file.");
      System.exit(-1);
    }
    final String HOST = args[0];
    final String AGENT_EMAIL = args[1];
    final String AGENT_PRIV_KEY_ID = args[2];
    final String AGENT_PRIV_KEY_FILE = args[3];

    // The private key should start with "-----BEGIN RSA PRIVATE KEY-----" and
    // end with "-----END RSA PRIVATE KEY-----". In between, there should be newline-delimtted
    // strings of characters.
    String privateKey;
    try {
      privateKey = new String(Files.readAllBytes(Paths.get(AGENT_PRIV_KEY_FILE)));
    } catch (IOException e) {
      throw new RuntimeException("Private key could not be read.", e);
    }

    // Sets up the channel using the two signed JWTs for RPCs to the NBI.
    ManagedChannel channel =
        newChannelBuilderForAddress(
                HOST,
                PORT,
                CompositeChannelCredentials.create(
                    TlsChannelCredentials.create(),
                    SpacetimeCallCredentials.createFromPrivateKey(
                        HOST, AGENT_EMAIL, AGENT_PRIV_KEY_ID, privateKey)))
            .enableRetry()
            .build();

    // Sets up a stub to invoke RPCs against the NBI's NetOps service.
    NetOpsGrpc.NetOpsBlockingStub stub = NetOpsGrpc.newBlockingStub(channel);

    // This stub can now be used to call any method in the NetOps service.
    ListEntitiesRequest request =
        ListEntitiesRequest.newBuilder().setType(EntityType.PLATFORM_DEFINITION.name()).build();
    ListEntitiesResponse entities = stub.listEntities(request);
    System.out.println("ListEntitiesResponse received: \n" + entities.toString());
  }
}
